kube/src/api/remote_command.rs

```diff
@@ -0,0 +1,273 @@
+};
+
+use k8s_openapi::apimachinery::pkg::apis::meta::v1::Status;
+
+use async_tungstenite::{
+    tokio::ConnectStream,
+    tungstenite::{self as ws, Message},
+    WebSocketStream,
+};
+use futures::{future::Either, SinkExt, Stream, StreamExt, TryStreamExt};
+use futures_util::future::select;
+use tokio::io::{AsyncRead, AsyncWrite, AsyncWriteExt, DuplexStream};
+use tokio_util::codec;
+
+// Internal state of an attached process
+struct AttachedProcessState {
+    waker: Option<Waker>,
+    finished: bool,
+    status: Option<Status>,
+    stdin_writer: Option<DuplexStream>,
+    stdout_reader: Option<DuplexStream>,
+    stderr_reader: Option<DuplexStream>,
+}
+
+/// Represents an attached process in a container for `attach` and `exec`.
+///
+/// Provides access to stdin/stdout/stderr if attached.
+/// Resolves when the connection terminates with an optional [`Status`].
+///
+/// [`Status`]: k8s_openapi::apimachinery::pkg::apis::meta::v1::Status
+pub struct AttachedProcess {
+    has_stdin: bool,
+    has_stdout: bool,
+    has_stderr: bool,
+    state: Arc<Mutex<AttachedProcessState>>,
+}
+
+impl AttachedProcess {
+    pub(crate) fn new(
+        stream: WebSocketStream<ConnectStream>,
+        stdin: bool,
+        stdout: bool,
+        stderr: bool,
+    ) -> Self {
+        // To simplify the implementation, always create a pipe for stdin.
+        // The caller does not have access to it unless they had requested.
+        // REVIEW Make internal buffer size configurable?
+        let (stdin_writer, stdin_reader) = tokio::io::duplex(1024);
+        let (stdout_writer, stdout_reader) = if stdout {
+            let (w, r) = tokio::io::duplex(1024);
+            (Some(w), Some(r))
+        } else {
+            (None, None)
+        };
+        let (stderr_writer, stderr_reader) = if stderr {
+            let (w, r) = tokio::io::duplex(1024);
+            (Some(w), Some(r))
+        } else {
+            (None, None)
+        };
+
+        let state = Arc::new(Mutex::new(AttachedProcessState {
+            waker: None,
+            finished: false,
+            status: None,
+            stdin_writer: Some(stdin_writer),
+            stdout_reader,
+            stderr_reader,
+        }));
+        let shared_state = state.clone();
+        tokio::spawn(async move {
+            let status = start(stream, stdin_reader, stdout_writer, stderr_writer)
+                .await
+                .unwrap();
+
+            let mut shared = shared_state.lock().unwrap();
+            shared.finished = true;
+            shared.status = status;
+            if let Some(waker) = shared.waker.take() {
+                waker.wake()
+            }
+        });
+
+        AttachedProcess {
+            has_stdin: stdin,
+            has_stdout: stdout,
+            has_stderr: stderr,
+            state,
+        }
+    }
+
+    /// Async writer to write to stdin of the attached process.
+    pub fn stdin(&mut self) -> Option<impl AsyncWrite + Unpin> {
```

Comment:
I think it's better for this to return `futures::Sink`, but having some trouble converting.

Something like:
```rust
fn into_bytes_sink<W: AsyncWrite>(writer: W) -> impl Sink<&[u8]> {
    codec::FramedWrite::new(writer, codec::BytesCodec::new())
        .with(|s: &[u8]| async { Ok::<_, std::io::Error>(Bytes::from(s)) })
}
```
But need to specify lifetimes.

Reply:
I decided to return `reader` and `writer` instead of `stream` and `sink` because users can easily convert them.